### ISSUE TITLE
Question on Task 5

### ISSUE THREAD

**Issue by ingomueller-net** [root-project/opendata-benchmarks#1]
Task 5 contains the following [two lines](https://github.com/stwunsch/opendata-benchmarks/blob/master/tasks/5/rdataframe.C#L7), but I am struggling to match them to the task description:

```cpp
    const auto c = ROOT::VecOps::Combinations(pt.size(), 2);
    for (auto i = 0; i < pt.size(); i++) {
```

I suppose that finding "an opposite-sign muon pair" means that we should compare any two muons pairwise, right? I `VecOps::Combinations` seem to produce indices of all pairwise combinations. However, the loop doesn't iterate over all of these combinations; it ends at `pt.size()`. Is this a mistake? Shouldn't it be `c[0].size()` or similar?

I have zero knowledge in Go, so I might be reading the [Go implementation](https://github.com/go-hep/examples/blob/master/groot/bench-opendata/05-basic.go#L72) incorrectly, but that one seems to iterate over all combinations as well:

```go
		combs := combin.Combinations(nmuons, 2)
		for _, c := range combs {
```

**Comment by ingomueller-net** [comment]
Actually, I think that the first of the two lines contains another bug (or something I don't understand): The second parameter of `Combinations` should be `pt.size()` also, no? As I understand that function, it essentially produces the indices of a nested for loop. Here, it should be used to test all pairwise combinations -- of indices into `pt`, no?

We probably only want to have only one out of `(x,y)` and `(y,x)`, so we should test for that too. If I am correct, the whole function should look like this:

```cpp
auto compute_dimuon_masses(Vec<float> pt, Vec<float> eta, Vec<float> phi, Vec<float> mass, Vec<int> charge)
{
    ROOT::RVec<float> masses;
    const auto c = ROOT::VecOps::Combinations(pt.size(), pt.size()); // take pairwise indices of pt and pt
    for (auto i = 0; i < c[0].size(); i++) {  // iterate over all index pairs ...
        const auto i1 = c[0][i];
        const auto i2 = c[1][i];
        if (i2 < i1) continue;  // ... but ignore duplicates
        if (charge[i1] == charge[i2]) continue;
        const FourVector p1(pt[i1], eta[i1], phi[i1], mass[i1]);
        const FourVector p2(pt[i2], eta[i2], phi[i2], mass[i2]);
        masses.push_back((p1 + p2).mass());
    }
    return masses;
};
```

Can you confirm or point me to where I am wrong?

**Comment by ingomueller-net** [comment]
OK, I think I found it. I think that the wrong overload is called. What you wanted to call is the [`Combinations(RVec, size_t)`](https://root.cern/doc/master/namespaceROOT_1_1VecOps.html#abbbf82257156b750c7b9bdd39a53ee33) overload, which produces all unique combinations of `2` indices inside of the vector. Then, the two lines should look like this:

```cpp
    const auto c = ROOT::VecOps::Combinations(pt, 2);  // no call to .size()
    for (auto i = 0; i < pt.size(); i++) {
```

Without having understood the Go implementation: that could be wrong, too.

@jpivarski: Can you confirm?

**Comment by jpivarski** [comment]
I don't know the details of the Go implementation but I bet @sbinet does.

**Comment by sbinet** [comment]
thanks.

but the [gonum/stat/combin.Combinations](https://godoc.org/gonum.org/v1/gonum/stat/combin#Combinations) function generates the unique pairs:
- https://play.golang.org/p/-JTlHJRkgpd
```
Generate list:
0 [0 1]
1 [0 2]
2 [0 3]
3 [0 4]
4 [1 2]
5 [1 3]
6 [1 4]
7 [2 3]
8 [2 4]
9 [3 4]
```

**Comment by ingomueller-net** [comment]
OK, thanks a lot, I see. `Combination(3 , 3)` in C++ and Go don't seem to mean the same thing. The Go implementation doesn't have the problem that the one in C++ has.

**Comment by stwunsch** [comment]
Thanks for finding this, that's great! I'll fix this with your commit in #4
